### ISSUE TITLE
a `libclang` shared library is not loaded on this thread

### ISSUE THREAD

**Issue by albx79** [twistedfall/opencv-rust#462]
Before reporting an issue please first check the

1. Operating system: Fedora 38
3. The way you installed OpenCV: package
4. OpenCV version: opencv-devel-4.7.0-2.fc38.x86_64
5. rustc version: rustc 1.69.0 (84c898d65 2023-04-16)
6. Attach the full output of the following command from your project directory:
   ```shell script
   RUST_BACKTRACE=full cargo build -vv 
   ```

**Comment by albx79** [comment]
[cargo.stderr.txt](https://github.com/twistedfall/opencv-rust/files/11422425/cargo.stderr.txt)
[cargo.stdout.txt](https://github.com/twistedfall/opencv-rust/files/11422426/cargo.stdout.txt)

**Comment by albx79** [comment]
I've just tried with an alpine docker image and I get the same error

**Comment by twistedfall** [comment]
Looks like bindgen now defaults to non-runtime clang. Looks like it's time to bring that feature back.

**Comment by albx79** [comment]
Thanks. In the meantime, as a workaround, would it work if I revert to an
older version (one using an older bindgen)?

Il giorno mar 9 mag 2023 alle ore 18:35 twistedfall <
***@***.***> ha scritto:

**Comment by twistedfall** [comment]
The best solution would be to contact `libspa-sys` developers so that they would introduce the `runtime` feature thus enabling `bindgen` with `runtime` too.

**Comment by twistedfall** [comment]
Can you please check out the version 0.81.3? This issue should be fixed there with no changes to external crates

**Comment by albx79** [comment]
Hi! Thanks, but unfortunately I'm still getting the same error (both on my Fedora, and on the Alpine container).

I'm attaching the new `cargo build -vv` output.
[fedora.stderr.txt](https://github.com/twistedfall/opencv-rust/files/11467528/fedora.stderr.txt)
[fedora.stdout.txt](https://github.com/twistedfall/opencv-rust/files/11467530/fedora.stdout.txt)

**Comment by zhouzq-thu** [comment]
## Same issue on macOS
- rustc 1.69.0 (84c898d65 2023-04-16)
- opencv 4.7.0
- opencv-rust 0.81.3

```
[opencv 0.81.3] dyld[86631]: Library not loaded: '@rpath/libclang.dylib'
[opencv 0.81.3]   Referenced from: '/Users/zhou/Desktop/rust/opencv_example/target/debug/build/opencv-85ae2efac4e5d0d4/build-script-build'
[opencv 0.81.3]   Reason: tried: '/Users/zhou/Desktop/rust/opencv_example/target/debug/deps/libclang.dylib' (no such file), '/Users/zhou/Desktop/rust/opencv_example/target/debug/libclang.dylib' (no such file), '/Users/zhou/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libclang.dylib' (no such file), '/Users/zhou/.rustup/toolchains/stable-x86_64-apple-darwin/lib/libclang.dylib' (no such file), '/Users/zhou/lib/libclang.dylib' (no such file), '/usr/local/lib/libclang.dylib' (no such file), '/usr/lib/libclang.dylib' (no such file)
error: failed to run custom build command for `opencv v0.81.3`

Caused by:
  process didn't exit successfully: `/Users/zhou/Desktop/rust/opencv_example/target/debug/build/opencv-85ae2efac4e5d0d4/build-script-build` (signal: 6, SIGABRT: process abort signal)
  --- stderr
  dyld[86631]: Library not loaded: '@rpath/libclang.dylib'
    Referenced from: '/Users/zhou/Desktop/rust/opencv_example/target/debug/build/opencv-85ae2efac4e5d0d4/build-script-build'
    Reason: tried: '/Users/zhou/Desktop/rust/opencv_example/target/debug/deps/libclang.dylib' (no such file), '/Users/zhou/Desktop/rust/opencv_example/target/debug/libclang.dylib' (no such file), '/Users/zhou/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libclang.dylib' (no such file), '/Users/zhou/.rustup/toolchains/stable-x86_64-apple-darwin/lib/libclang.dylib' (no such file), '/Users/zhou/lib/libclang.dylib' (no such file), '/usr/local/lib/libclang.dylib' (no such file), '/usr/lib/libclang.dylib' (no such file)
```

## However, 0.80.0 can be built correctly.

**Comment by twistedfall** [comment]
@albx79 Those logs don't look like they were produced by `cargo build -vv` or they are truncated, they should contain some way more debug information. Can you double-check that?

@zhouzq-thu Can you please attach the full build log as produced by `cargo build -vv`?

**Comment by zhouzq-thu** [comment]
@twistedfall 
[output.txt](https://github.com/twistedfall/opencv-rust/files/11469231/output.txt)

**Comment by albx79** [comment]
`RUST_BACKTRACE=1 cargo build -vv > fedora.stdout.txt 2> fedora.stderr.txt`

[fedora.stderr.txt](https://github.com/twistedfall/opencv-rust/files/11469988/fedora.stderr.txt)
[fedora.stdout.txt](https://github.com/twistedfall/opencv-rust/files/11469989/fedora.stdout.txt)

**Comment by twistedfall** [comment]
@zhouzq-thu Can you please follow item 7 from the [Troubleshooting section](https://github.com/twistedfall/opencv-rust/blob/master/README.md#troubleshooting)?

**Comment by twistedfall** [comment]
@albx79 Can you please try using the latest version 0.81.3 which should fix this particular problem.

**Comment by zhouzq-thu** [comment]
@twistedfall I tried, but still failed to build.

**Comment by hsyan2008** [comment]
> @zhouzq-thu Can you please follow item 7 from the [Troubleshooting section](https://github.com/twistedfall/opencv-rust/blob/master/README.md#troubleshooting)?

with opencv-rust 0.81.3
I build musl success on Mac

**Comment by twistedfall** [comment]
@zhouzq-thu Can you please double-check if you set up the environment variable correctly because the error message doesn't seem to mention any directory that would belong to Xcode:
> [opencv 0.81.3]   Reason: tried: '/Users/zhou/Desktop/rust/opencv_sample/target/debug/deps/libclang.dylib' (no such file), '/Users/zhou/Desktop/rust/opencv_sample/target/debug/libclang.dylib' (no such file), '/Users/zhou/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libclang.dylib' (no such file), '/Users/zhou/.rustup/toolchains/stable-x86_64-apple-darwin/lib/libclang.dylib' (no such file), '/Users/zhou/lib/libclang.dylib' (no such file), '/usr/local/lib/libclang.dylib' (no such file), '/usr/lib/libclang.dylib' (no such file)

**Comment by AdrianEddy** [comment]
I have the same issue with 0.81.3 and 0.81.4 on windows. 
0.81.2 builds without issues
[0.81.4.txt](https://github.com/twistedfall/opencv-rust/files/11486088/0.81.4.txt)
[0.81.2.txt](https://github.com/twistedfall/opencv-rust/files/11486089/0.81.2.txt)

**Comment by zhouzq-thu** [comment]
@twistedfall @hsyan2008 I found that it is OK if no other crates in the project, but failed if some crates in it.

Try this:

[opencv-compile-test.zip](https://github.com/twistedfall/opencv-rust/files/11486122/opencv-compile-test.zip)

With 0.81.4,
```shell
export DYLD_FALLBACK_LIBRARY_PATH="$(xcode-select --print-path)/Toolchains/XcodeDefault.xctoolchain/usr/lib/" && cargo build --all --target=x86_64-apple-darwin
```
generates:
```
  === Detected OpenCV module header dir at: /usr/local/opt/opencv/include/opencv4/opencv2
  === Found OpenCV version: 4.7.0 in headers located at: /usr/local/opt/opencv/include/opencv4
  === Generating code in: /Users/zhou/Desktop/rust/opencv-compile-test/target/x86_64-apple-darwin/debug/build/opencv-c8232f0434341892/out
  === Placing generated bindings into: /Users/zhou/Desktop/rust/opencv-compile-test/target/x86_64-apple-darwin/debug/build/opencv-c8232f0434341892/out/opencv
  === Using OpenCV headers from: /usr/local/opt/opencv/include/opencv4
  thread 'main' panicked at 'a `libclang` shared library is not loaded on this thread', /Users/zhou/.cargo/registry/src/github.com-1ecc6299db9ec823/clang-sys-1.6.1/src/lib.rs:1735:1
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

0.80.0 can be built correctly.

**Comment by hsyan2008** [comment]
> @twistedfall @hsyan2008 I found that it is OK if no other crates in the project, but failed if some crates in it.
> 
> Try this:
> 
> [opencv-compile-test.zip](https://github.com/twistedfall/opencv-rust/files/11486122/opencv-compile-test.zip)
> 
> With 0.81.4,
> 
> ```shell
> export DYLD_FALLBACK_LIBRARY_PATH="$(xcode-select --print-path)/Toolchains/XcodeDefault.xctoolchain/usr/lib/" && cargo build --all --target=x86_64-apple-darwin
> ```
> 
> generates:
> 
> ```
>   === Detected OpenCV module header dir at: /usr/local/opt/opencv/include/opencv4/opencv2
>   === Found OpenCV version: 4.7.0 in headers located at: /usr/local/opt/opencv/include/opencv4
>   === Generating code in: /Users/zhou/Desktop/rust/opencv-compile-test/target/x86_64-apple-darwin/debug/build/opencv-c8232f0434341892/out
>   === Placing generated bindings into: /Users/zhou/Desktop/rust/opencv-compile-test/target/x86_64-apple-darwin/debug/build/opencv-c8232f0434341892/out/opencv
>   === Using OpenCV headers from: /usr/local/opt/opencv/include/opencv4
>   thread 'main' panicked at 'a `libclang` shared library is not loaded on this thread', /Users/zhou/.cargo/registry/src/github.com-1ecc6299db9ec823/clang-sys-1.6.1/src/lib.rs:1735:1
>   note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> ```
> 
> 0.80.0 can be built correctly.

@zhouzq-thu 

In my Cargo.toml like this
```
log = "0.4.17"
tokio = { version = "1.26", features = ["full"]  }
anyhow = "1.0.70"
thirtyfour = { version = "0.31.0" }
toml = "0.7.3"
serde = "1.0.160"
tokio-cron-scheduler = { version = "0.9.4", features = ["signal"] }
rbatis = { version = "3.1", default-features = false, features = ["runtime-tokio-rustls","mysql"]}
rbson = { version = "2.0", features = ["uuid-0_8", "chrono-0_4"] }
futures = "0.3.28"
serde_json = "1.0.96"
clap = "4.2.4"

axum = "0.5.6"
which = "4.4.0"
chromiumoxide = "0.5.0"
base64 = "0.21.0"
serde_repr = "0.1.12"
opencv = "0.81.4"

```
and build by
```
#opencv library in /usr/local/lib/
export DYLD_FALLBACK_LIBRARY_PATH="/usr/local/lib/:/Library/Developer/CommandLineTools/usr/lib/"
cargo build --target=x86_64-apple-darwin
```
```
cargo build --release --target x86_64-unknown-linux-musl
```
all above is success

Can you try using
```
cargo build --release --target x86_64-unknown-linux-musl
```

**Comment by twistedfall** [comment]
@AdrianEddy Can you please try version 0.81.5 with `clang-runtime` flag enabled?

**Comment by AdrianEddy** [comment]
0.81.5 works with "clang-runtime" feature enabled, thanks

**Comment by twistedfall** [comment]
@zhouzq-thu The same solution should also work for you: make sure to use version 0.81.5 and enable the new `clang-runtime` feature of the `opencv` crate.

**Comment by zhouzq-thu** [comment]
> @zhouzq-thu The same solution should also work for you: make sure to use version 0.81.5 and enable the new `clang-runtime` feature of the `opencv` crate.

0.81.5 works with "clang-runtime" feature enabled on macOS. Thanks.

**Comment by jaques-sam** [comment]
Mmm, I still get the error on debian10 (with opencv version 3.2.0) even with the `clang-runtime` feature enabled. As a workaround I'm using 0.75 where the issue is not appearing.
UPDATE: I just checked, for me the **last working version is 0.77** before https://github.com/twistedfall/opencv-rust/issues/455 occured.

**Comment by albx79** [comment]
On Fedora, I'm still getting the error with the latest 0.82 (but also with the old 0.77). Same story on Alpine.

@twistedfall , if you want you can clone my project (https://github.com/albx79/ademas, technically it's unlicensed but you can take this as permission to use), and use the dockerfile there to test a Linux build.

**Comment by twistedfall** [comment]
@albx79 That project is private and I can't access it unfortunately.
@jaques-sam Can you please provide a full build log as produced by `cargo build -vv`?

**Comment by twistedfall** [comment]
@jaques-sam Ah, wait, I've missed the fact that you're using OpenCV 3.2, that version is no longer supported so you should probably stick with whatever crate version works for you until you are able to switch to OpenCV 3.4 or 4.x

**Comment by albx79** [comment]
> @albx79 That project is private and I can't access it unfortunately.

Sorry @twistedfall , I had forgotten about that. I've added you as collaborator. If it still doesn't work, I'll make it public for a while.

**Comment by twistedfall** [comment]
I can see the problem now. The general idea is that all crates using `clang-sys` must either all enable its `runtime` feature or all not. It won't work if some are enabling it and some are not (at least with rust `edition` before 2021 with `resolver` before 2).

In your case there are 3 dependencies that are using `clang-sys`: `opencv`, `v4l` and `libspa`. The latter two are using it via `bindgen`. The `bindgen` crate enables the `runtime` feature of `clang-sys` by default, but allows opting out of it.

The problem is that `v4l` uses `bindgen` with [default features](https://github.com/raymanfx/libv4l-rs/blob/master/v4l2-sys/Cargo.toml#L11) (=`runtime` enabled), but `libspa` uses it with default features [disabled](https://gitlab.freedesktop.org/pipewire/pipewire-rs/-/blob/main/libspa-sys/Cargo.toml#L19) (=`runtime` disabled). Neither of those allow you to control it (with `opencv` you can optionally enable `runtime` by using `clang-runtime` feature).

So long story short, your build will fail even without `opencv` and to solve it you should either enabled the default features with `bindgen` in `libspa-sys` or disable them in `v4l2-sys` and adjust the `clang-runtime` feature for `opencv` accordingly. That can be done by either creating issues or PRs in the corresponding repositories or forking the repos and doing the necessary changes yourself.

**Comment by albx79** [comment]
Thanks @twistedfall for the thorough analysis. I think my best chance at this point is to drop some dependencies, starting from V4L, and see if I can make it work from there.

**Comment by twistedfall** [comment]
> drop some dependencies, starting from V4L, and see if I can make it work from there.

OpenCV has functions to access the webcam so you probably can use that instead.

**Comment by aaqidmasoodi** [comment]
can someone help with this. it not able to find something not sure how to fix:

=== Successfully probed using: environment
  === OpenCV library configuration: Library {
      include_paths: [
          "/opt/homebrew/Cellar/opencv/4.8.0_4/include",
      ],
      version: Version {
          major: 0,
          minor: 0,
          patch: 0,
      },
      cargo_metadata: [
          "cargo:rustc-link-search=/opt/homebrew/Cellar/opencv/4.8.0_4/share/opencv4",
          "cargo:rustc-link-search=framework=/opt/homebrew/Cellar/opencv/4.8.0_4/share/opencv4",
          "cargo:rustc-link-lib=",
      ],
  }
  thread 'main' panicked at 'Unsupported OpenCV version: 0.0.0, must be from 3.2, 3.4 or 4.x branch', /Users/aaqidmasoodi/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencv-0.84.5/build.rs:354:9
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

**Comment by kevleyski** [comment]
Ah yep (macOS)
opencv = { version = "0.85.3", features = ["clang-runtime"] }

Maybe there is a way to detect target, or flag the error with some helper somehow, README.md warning? 
Either way working ok, thanks

**Comment by twistedfall** [comment]
Since version 0.86.1 you should be getting a more friendly message when libclang loading mismatch is detected: https://github.com/twistedfall/opencv-rust/blob/730cc64955ab4ad5286bee891e795a1322b053c0/build/generator.rs#L74-L77. Thanks for the suggestion!

**Comment by kevleyski** [comment]
Great thanks!
